### PR TITLE
8268821: Split systemDictionaryShared.cpp

### PR DESCRIPTION
**Pull request by yminqi**
Hi, Please review
systemDictionaryShared becomes fatter and fatter so it is time to split it into functional files. Moved security and jar operation related code into CDSProtectionDomain, and moved shared class info (DumpTime/RunTime) to sharedClassInfo.[ch]pp, also moved lambda proxy related to lambdaProxyClassInfo.hpp. This way systemDictionaryShared.cpp looks neat and light.

Tests: tier1,tier3,tier4

Thanks
Yumin

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8268821](https://bugs.openjdk.java.net/browse/JDK-8268821): Split systemDictionaryShared.cpp

### Reviewers
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to e3cd258f0777fc4a2aa5c5dde443acc8a3c9db92
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**)
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4568/head:pull/4568` \
`$ git checkout pull/4568`

Update a local copy of the PR: \
`$ git checkout pull/4568` \
`$ git pull https://git.openjdk.java.net/jdk pull/4568/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4568`

View PR using the GUI difftool: \
`$ git pr show -t 4568`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4568.diff">https://git.openjdk.java.net/jdk/pull/4568.diff</a>

</details>

### CODE REVIEW

**Review by erikj79 — APPROVED** [review]
Build changes look ok.

**Comment by yminqi** [comment]
@erikj79 Thanks for review!

**Comment by yminqi** [comment]
> _Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [build-dev](mailto:build-dev@mail.openjdk.java.net):_
> 
> Hi Yumin,
> 
> On 23/06/2021 4:19 pm, Yumin Qi wrote:
> 
> > Hi, Please review
> > systemDictionaryShared becomes fatter and fatter so it is time to split it into functional files. Moved security and jar operation related code into CDSProtectionDomain, and moved shared class info (DumpTime/RunTime) to sharedClassInfo.[ch]pp, also moved lambda proxy related to lambdaProxyClassInfo.hpp. This way systemDictionaryShared.cpp looks neat and light.
> 
> I'm not really seeing a consistent or recognisable naming pattern here.
> We seem to have a mix of:
> 
> - cds/foo.cpp
> - cds/fooShared.cpp
> - cds/sharedFoo.cpp
> 
> Can we establish a simple naming scheme here?
> 
> Thanks,
> David

Thanks David. I was thinking of that too. The best practice is for a class Foo we have foo.hpp for definition and foo.cpp for implementation. Here indeed exists non-consistency that I put DumpTime/RunTtime in a single file. Let me double check and update.
Yumin

File: src/hotspot/share/cds/cdsProtectionDomain.cpp [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+#include "precompiled.hpp"
+#include "cds/cdsProtectionDomain.hpp"
+#include "classfile/javaClasses.hpp"
+#include "classfile/classLoader.hpp"
+#include "classfile/classLoaderExt.hpp"
+#include "classfile/moduleEntry.hpp"
+#include "classfile/systemDictionaryShared.hpp"
+#include "classfile/vmClasses.hpp"
+#include "classfile/vmSymbols.hpp"
+#include "classfile/symbolTable.hpp"
```
Comment (calvinccheung) [inline src/hotspot/share/cds/cdsProtectionDomain.cpp:34]: Please move line 34 before line 31.

File: src/hotspot/share/cds/dumpTimeSharedClassInfo.hpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+ * questions.
+ *
+ */
+
+#ifndef SHARED_CDS_DUMPTIMESHAREDCLASSINFO_HPP
+#define SHARED_CDS_DUMPTIMESHAREDCLASSINFO_HPP
+#include "classfile/compactHashtable.hpp"
+#include "cds/archiveBuilder.hpp"
+#include "cds/archiveUtils.hpp"
+#include "cds/metaspaceShared.hpp"
```
Comment (calvinccheung) [inline src/hotspot/share/cds/dumpTimeSharedClassInfo.hpp:31]: Please move line 28 after line 31.

File: src/hotspot/share/cds/metaspaceShared.cpp [hunk 3]
```diff
@@ -32,6 +32,7 @@
 #include "cds/heapShared.hpp"
 #include "cds/lambdaFormInvokers.hpp"
 #include "cds/metaspaceShared.hpp"
+#include "cds/cdsProtectionDomain.hpp"
```
Comment (calvinccheung) [inline src/hotspot/share/cds/metaspaceShared.cpp:35]: This line should be after line 27.

File: src/hotspot/share/classfile/systemDictionaryShared.hpp [hunk 4]
```diff
@@ -341,11 +285,10 @@ class SystemDictionaryShared: public SystemDictionary {
   static void print_on(outputStream* st) NOT_CDS_RETURN;
   static void print_shared_archive(outputStream* st, bool is_static = true) NOT_CDS_RETURN;
   static void print_table_statistics(outputStream* st) NOT_CDS_RETURN;
-  static bool empty_dumptime_table() NOT_CDS_RETURN_(true);
+  static bool is_dumptime_table_empty() NOT_CDS_RETURN_(true);
   static void start_dumping() NOT_CDS_RETURN;
-  static Handle create_jar_manifest(const char* man, size_t size, TRAPS) NOT_CDS_RETURN_(Handle());
   static bool is_supported_invokedynamic(BootstrapInfo* bsi) NOT_CDS_RETURN_(false);
-
+  static void dd_to_dump_time_lambda_proxy_class_dictionary(LambdaProxyClassKey key, InstanceKlass* proxy_klass);
```
Comment (calvinccheung) [inline src/hotspot/share/classfile/systemDictionaryShared.hpp:291]: I don't think this is being referenced.

**Review by calvinccheung — CHANGES_REQUESTED** [review]
Looks good. Few minor comments.

**Comment by yminqi** [comment]
> _Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [hotspot-runtime-dev](mailto:hotspot-runtime-dev@mail.openjdk.java.net):_
> 
> On 24/06/2021 2:23 am, Yumin Qi wrote:
> 
> > On Wed, 23 Jun 2021 06:08:41 GMT, Yumin Qi <minqi at openjdk.org> wrote:
> > > Hi, Please review
> > > systemDictionaryShared becomes fatter and fatter so it is time to split it into functional files. Moved security and jar operation related code into CDSProtectionDomain, and moved shared class info (DumpTime/RunTime) to sharedClassInfo.[ch]pp, also moved lambda proxy related to lambdaProxyClassInfo.hpp. This way systemDictionaryShared.cpp looks neat and light.
> > > Tests: tier1,tier3,tier4
> > > Thanks
> > > Yumin
> > 
> > 
> > > _Mailing list message from [David Holmes](mailto:david.holmes at oracle.com) on [build-dev](mailto:build-dev at mail.openjdk.java.net):_
> > > Hi Yumin,
> > > On 23/06/2021 4:19 pm, Yumin Qi wrote:
> > > > Hi, Please review
> > > > systemDictionaryShared becomes fatter and fatter so it is time to split it into functional files. Moved security and jar operation related code into CDSProtectionDomain, and moved shared class info (DumpTime/RunTime) to sharedClassInfo.[ch]pp, also moved lambda proxy related to lambdaProxyClassInfo.hpp. This way systemDictionaryShared.cpp looks neat and light.
> > > 
> > > 
> > > I'm not really seeing a consistent or recognisable naming pattern here.
> > > We seem to have a mix of:
> > > - cds/foo.cpp
> > > - cds/fooShared.cpp
> > > - cds/sharedFoo.cpp
> > > Can we establish a simple naming scheme here?
> > > Thanks,
> > > David
> > 
> > 
> > Thanks David. I was thinking of that too. The best practice is for a class Foo we have foo.hpp for definition and foo.cpp for implementation. Here indeed exists non-consistency that I put DumpTime/RunTtime in a single file. Let me double check and update.
> 
> That's not what I was saying. I'm talking about the names of the cpp
> file and whether they contain "shared" and whether it is a prefix or
> postfix. There doesn't seem to be a consistent naming scheme employed here.
That comes from day one.  The case class FooShared is like 
cds/fooShared.[ch]pp
Usually it is for a class with counterpart of a non-shared version, like Metaspace, SystemDictionary etc.
The classes [DumpTime/RunTime]SharedClassInfo are used for shared only, they don't have non-shared version. The "Shared" embedded just an indication of used in CDS, without it is OK i think.

**Comment by calvinccheung** [comment]
> > _Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [hotspot-runtime-dev](mailto:hotspot-runtime-dev@mail.openjdk.java.net):_
> > On 24/06/2021 2:23 am, Yumin Qi wrote:
> > > On Wed, 23 Jun 2021 06:08:41 GMT, Yumin Qi  wrote:
> > > > Hi, Please review
> > > > systemDictionaryShared becomes fatter and fatter so it is time to split it into functional files. Moved security and jar operation related code into CDSProtectionDomain, and moved shared class info (DumpTime/RunTime) to sharedClassInfo.[ch]pp, also moved lambda proxy related to lambdaProxyClassInfo.hpp. This way systemDictionaryShared.cpp looks neat and light.
> > > > Tests: tier1,tier3,tier4
> > > > Thanks
> > > > Yumin
> > > 
> > > 
> > > > _Mailing list message from [David Holmes](mailto:david.holmes at oracle.com) on [build-dev](mailto:build-dev at mail.openjdk.java.net):_
> > > > Hi Yumin,
> > > > On 23/06/2021 4:19 pm, Yumin Qi wrote:
> > > > > Hi, Please review
> > > > > systemDictionaryShared becomes fatter and fatter so it is time to split it into functional files. Moved security and jar operation related code into CDSProtectionDomain, and moved shared class info (DumpTime/RunTime) to sharedClassInfo.[ch]pp, also moved lambda proxy related to lambdaProxyClassInfo.hpp. This way systemDictionaryShared.cpp looks neat and light.
> > > > 
> > > > 
> > > > I'm not really seeing a consistent or recognisable naming pattern here.
> > > > We seem to have a mix of:
> > > > 
> > > > * cds/foo.cpp
> > > > * cds/fooShared.cpp
> > > > * cds/sharedFoo.cpp
> > > >   Can we establish a simple naming scheme here?
> > > >   Thanks,
> > > >   David
> > > 
> > > 
> > > Thanks David. I was thinking of that too. The best practice is for a class Foo we have foo.hpp for definition and foo.cpp for implementation. Here indeed exists non-consistency that I put DumpTime/RunTtime in a single file. Let me double check and update.
> > 
> > 
> > That's not what I was saying. I'm talking about the names of the cpp
> > file and whether they contain "shared" and whether it is a prefix or
> > postfix. There doesn't seem to be a consistent naming scheme employed here.
> 
> That comes from day one. The case class FooShared is like
> cds/fooShared.[ch]pp
> Usually it is for a class with counterpart of a non-shared version, like Metaspace, SystemDictionary etc.
> The classes [DumpTime/RunTime]SharedClassInfo are used for shared only, they don't have non-shared version. The "Shared" embedded just an indication of used in CDS, without it is OK i think.

I agree that it's fine to drop the word "Shared" from the [DumpTime/RunTime]SharedClassInfo so the the new files could be named [dump|run]TimeClassInfo.[c|h]pp instead.

File: src/hotspot/share/cds/lambdaProxyClassDictionary.hpp [hunk 5]
```diff
@@ -25,6 +25,7 @@
 #ifndef SHARED_CDS_LAMBDAPROXYCLASSINFO_HPP
 #define SHARED_CDS_LAMBDAPROXYCLASSINFO_HPP
 #include "cds/metaspaceShared.hpp"
+#include "classfile/javaClasses.hpp"
```
Comment (calvinccheung) [inline src/hotspot/share/cds/lambdaProxyClassDictionary.hpp:28]: Is this include necessary?

**Review by calvinccheung — APPROVED** [review]
Latest changes look good. One question.

File: src/hotspot/share/cds/lambdaProxyClassDictionary.hpp (hunk 5 above)
Comment (yminqi) [reply src/hotspot/share/cds/lambdaProxyClassDictionary.hpp:28]: It is needed, line 84:
 return java_lang_String::hash_code((const jbyte*)sym->bytes(), sym->utf8_length());

Comment (calvinccheung) [reply src/hotspot/share/cds/lambdaProxyClassDictionary.hpp:28]: I see.

**Review by iklam — APPROVED** [review]
This looks OK to me.

The naming of the various classes is still not consistent since the code has evolved for a long time. It was probably a mistake to add `Shared` as a classname suffix, but that was done since the beginning of CDS.

Maybe eventually we should rename all the classes with a CDS prefix (similar to the JFR code). That should be done in a separate PR, though.

**Comment by yminqi** [comment]
@iklam @calvinccheung @erikj79 
Thank you for review!
/integrate

**Pull request closed without merging**